### ISSUE TITLE
Add 'auto' mode

### ISSUE THREAD

**Issue by mleonhard** [pgbouncer/pgbouncer#467]
Suggested update to [docs](https://www.pgbouncer.org/config.html):
> `pool_mode` - Specifies when a server connection can be reused by other clients.
> - `auto`: Default. **NEW** 
>   - If client modifies session parameters, connection goes back to pool after client disconnects.
>   - If client starts a transaction, connection goes back to pool after transaction finishes.
>   - If client does not modify session or start a transaction, connection goes back to pool after each statement. 
> - `session`: Server is released back to pool after client disconnects. ~~Default.~~
> - `transaction`: Server is released back to pool after transaction finishes.
> - `statement`: Server is released back to pool after query finishes. Transactions spanning multiple statements are disallowed in this mode.

This new `auto` mode has several benefits:
1. One connection pool can service applications and database management tools.
2. Prevents users from unknowingly re-using sessions.  Examples:
   - #418 pgbouncer is reusing session connections in transaction mode without resetting them
   - #110 New server_reset_query_always default introduces non-deterministic behavior.
   - #452 Add a warning about tools such as pg_dump w.r.t. transaction pooling.
   - #396 handle all session modifying postgres statements like transactions
   - #369 Unable to PAUSE the server with pool_mode=session
   - #224 Connection read-only flag no reset
   - #203 PGBouncer swapping query parameters in txn pooling mode
   - #164 Wrong response with pgbouncer
   - [Prisma2 DigitalOcean Postgres + PGBouncer Issues](https://github.com/prisma/prisma2/issues/1638)
   - [CVE-2018-1058](https://wiki.postgresql.org/wiki/A_Guide_to_CVE-2018-1058%3A_Protect_Your_Search_Path)
   - [Consistent search_path between client and server](https://github.com/pgbouncer/pgbouncer/pull/73)
   - #242 Protocol level prepared statements
   - [Postgres - ERROR: prepared statement “S_1” already exists](https://stackoverflow.com/questions/7611926/postgres-error-prepared-statement-s-1-already-exists)

**Comment by Melkij** [comment]
Before update of the docs someone should implement such feature.

    select somefunc();

This query should be treated as session or statement pool? Its possible to `perform set_config(...`.

**Comment by petere** [comment]
Yeah, the problem with this proposal is that "If client modifies session parameters" is impossible to determine by PgBouncer.

**Comment by mleonhard** [comment]
@eulerto What happened to close this issue?  It seems like "auto" pool mode is not mentioned in the docs in the `master` branch.

https://github.com/pgbouncer/pgbouncer/blob/12f2c79e27ad91467250371be70d2832971564f6/doc/config.md?plain=1#L86-L100

**Comment by eulerto** [comment]
This issue was closed because you didn't provide answers to fix the design issues mentioned in the comments above. The PgBouncer purpose is to be a fast connection pooler, hence, it mustn't add any features that slowdown the performance (such as a query parser). From previous discussions, it is clear that PgBouncer won't implement a Postgres parser that would allow it to figure out if the SQL command modifies session parameters, hence, it is really hard to implement `pool_mode = auto`. It also introduces a non-deterministic behavior and even break applications (that expects `session` that always execute `server_reset_query` but it doesn't because the `auto` decides to execute in `statement` mode).

**Comment by mleonhard** [comment]
My own application broke in a surprising way when I deployed it on DigitalOcean's managed Postgres service which [uses PgBouncer](https://docs.digitalocean.com/products/databases/postgresql/how-to/manage-connection-pools/).  I spent a lot of time learning about PgBouncer and figuring out strange errors about missing prepared statements.  My application was not doing anything strange, just using a standard SQL library in the normal way.  The error happened because DigitalOcean runs PgBouncer in  `transaction` mode by default.  I think a lot of other application developers will hit this same problem.

Adding a new mode will not break any existing applications.

Changing the default mode to `auto` could break some applications.  It can also make PgBouncer work flawlessly for most users.  It's a tradeoff.  Even if we decide not to change the default, the `auto` mode is still worth implementing.  Hosting companies can make `auto` mode the default for their users.   Users can switch to it and solve their session interference problems.  This can save a lot of users a lot of trouble and make PgBouncer more useful and valuable.

> it is really hard to implement `pool_mode = auto`

This is a feature request.  Maybe someone will come along who wants to do that hard work?  If we're open to someone doing the work, then let's leave this issue open.
